### ISSUE TITLE
Possible issue with testcase: SmbdDataTransfer_IdleConnection_Server

### ISSUE THREAD

**Issue by vinitagnihotri** [microsoft/WindowsProtocolTestSuites#207]
We are running SMBD capable server in centos environment and using Windows 2012R2 as smbclient. Running MSFT-SMBD test suite 3.18.6.0

Test case SmbdDataTransfer_IdleConnection_Server fails with following error message:
```

ErrorMessage
Assert.AreEqual failed. Expected: , Actual: . Flags is NONE

===== Stack Trace =====

at Microsoft.Protocol.TestSuites.Smbd.TestSuite.SmbdDataTransfer.SmbdDataTransfer_IdleConnection_Server() in e:\Jenkins\workspace\CodeSign-SMBD-Test\TestSuite\TestSuites\MS-SMBD\src\TestSuite\SMBD\SmbdDataTransfer.cs:line 1206

=== End of Stack Trace ===

```
Why it fails even though Expected value is equal to Actual value?

**Comment by huiren-jiang** [comment]
Hi vinitagnihotri,

Thanks for reporting this issue!

1. The test case failed due to flags is NONE instead of SMB_DIRECT_RESPONSE_REQUESTED. This case expects server will send SMBD Data Transfer message with SMB_DIRECT_RESPONSE_REQUESTED flag when Idle Connection Timer is expired. Could you please check whether your server set the flag as expected?

1. The reason why the log was confusing in PTM is due to some issue of PTF. We will fix it later and notify you when it is available. Before that, you could use batch to launch the test and the output in .trx should be clear as below.

```
===========ErrorMessage==============
Assert.AreEqual failed. Expected: <SMB_DIRECT_RESPONSE_REQUESTED (Microsoft.Protocols.TestTools.StackSdk.FileAccessService.Smbd.SmbdDataTransfer_Flags)>, Actual: <NONE (Microsoft.Protocols.TestTools.StackSdk.FileAccessService.Smbd.SmbdDataTransfer_Flags)>. Flags is NONE
===== Stack Trace ===== 
```
 

Best Regards,

Windows Protocol Test Suites Development Team

**Comment by vinitagnihotri** [comment]
Hi,
I believe this is what happening:
1. Server sends out SMB_DIRECT_RESPONSE_REQUESTED on 120 seconds mark and resets timer for 5 seconds
2.test suite is expecting another message for timeout i.e. on 125 sec mark, however our server side logs shows test suite is sending out message with SMB_DIRECT_RESPONSE_REQUESTED bit on, which results into server sending out message earlier than connection timeout with flags as 0 as timer is still not expired (its at 124 sec mark)
Please find logs attached for this as well.

Specs mentions we should reply with SMB_DIRECT_RESPONSE_REQUESTED if we receive message with SMB_DIRECT_RESPONSE_REQUESTED bit set, but this does not seems to be valid while running testcases, as I have observed if we send out message with SMB_DIRECT_RESPONSE_REQUESTED without timer expiry test fails.
[Idleconnection_server_log.txt](https://github.com/microsoft/WindowsProtocolTestSuites/files/3723350/Idleconnection_server_log.txt)

**Comment by vinitagnihotri** [comment]
These are messages from our server logs, 120 sec timer has expired on 12:31:30, and we can see a incoming message on 12:31:34 with SMB_DIRECT_RESPONSE_REQUESTED bit on.

[2019-10-14 12:31:30.282]::[RDMA]: Idle timer expired for conn 0x7ff26c002d40 
[2019-10-14 12:31:30.283]::[RDMA]: sent SMB_DIRECT_RESPONSE_REQUESTED to conn 0x7ff26c002d40    
[2019-10-14 12:31:30.283]::[RDMA]: OUT connection=0x7ff26c002d40 credit:255:60
[2019-10-14 12:31:34.318]::[RDMA]: RDMA DATA transfer message received: id=0x7ff26c002d40
[2019-10-14 12:31:34.318]::[RDMA]: conn 0x7ff26c002d40 got SMB_DIRECT_RESPONSE_REQUESTED

**Comment by huiren-jiang** [comment]
Hi vinitagnihotri,

The test case is to verify the below server requirements and it is consistent with Windows Server:

- When client does not send any packet to server in 120 seconds, server is expected to send a keep-alive test packet to client with SMB_DIRECT_RESPONSE_REQUESTED set.

- Client has to reply a packet in 5 seconds, or server could terminate the RDMA connection. This is what `SmbdDataTransfer_IdleConnection_Server_Timeout` verifies.

- After client replies the keep-alive test with SMB_DIRECT_RESPONSE_REQUESTED set, server should also reply to indicate alive but without  SMB_DIRECT_RESPONSE_REQUESTED to avoid infinite loop. Since server should reset the 120 seconds idle timer, client will expect another keep-alive test packet afterwards. This is the case `SmbdDataTransfer_IdleConnection_Server`.

The corresponding parts in specification might not be that straightforward. I hope my explanation answered how test cases expect timer work. 

Best Regards,

Windows Protocol Test Suites Development Team

**Comment by vinitagnihotri** [comment]
Understood, seems like its bit grey area. I believe we are ok here. 
I am closing this issue.

Thanks and Regards.

**Comment by JiajunW** [comment]
For the first issue that actual and expected values are not showing, this has been fixed in microsoft/ProtocolTestFramework@eab8128.

Download [HtmlTestLogger.zip](https://github.com/microsoft/WindowsProtocolTestSuites/files/3737158/HtmlTestLogger.zip), extract the zip file, and you will get two files, *Microsoft.Protocols.TestTools.Html.TestLogger.dll* and *Microsoft.Protocols.TestTools.Html.TestLogger.pdb*. Go to <code>C:\Program Files (x86)\Microsoft Visual Studio\\<strong>2019</strong>\\<strong>Community</strong>\Common7\IDE\Extensions\TestPlatform\Extensions</code>, replace the .dll and .pdb file with the new one. Re-run the test cases in PTM, you will see the correct testing log.

*Note*: The above location is for Visual Studio 2019 community, if you are using a different version of Visual Studio, please modify the corresponding part.

Best Regards,
Windows Protocol Test Suites Development Team
